core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java

```diff
@@ -259,18 +261,35 @@ public AnnotatedServiceBindingBuilder accessLogWriter(AccessLogWriter accessLogW
      * @return {@link ServerBuilder} to continue building {@link Server}
      */
     public ServerBuilder build(Object service) {
+        requireNonNull(service, "service");
+        this.service = service;
+        serverBuilder.annotatedServiceBindingBuilder(this);
+        return serverBuilder;
+    }
+
+    /**
+     * Applies the {@link ServiceConfigBuilder} created with the configured
+     * {@link AnnotatedHttpServiceExtensions} to the {@link ServerBuilder}.
+     *
+     * @param extensions the {@link AnnotatedHttpServiceExtensions} at the server level.
+     */
+    void applyToServiceConfigBuilder(AnnotatedHttpServiceExtensions extensions) {
+        final List<ExceptionHandlerFunction> exceptionHandlerFunctions =
+                exceptionHandlerFunctionBuilder.addAll(extensions.exceptionHandlers()).build();
+        final List<RequestConverterFunction> requestConverterFunctions =
+                requestConverterFunctionBuilder.addAll(extensions.requestConverters()).build();
+        final List<ResponseConverterFunction> responseConverterFunctions =
+                responseConverterFunctionBuilder.addAll(extensions.responseConverters()).build();
+
         final List<AnnotatedHttpServiceElement> elements =
-                AnnotatedHttpServiceFactory.find(pathPrefix, service,
-                                                 exceptionHandlerFunctionBuilder.build(),
-                                                 requestConverterFunctionBuilder.build(),
-                                                 responseConverterFunctionBuilder.build());
+                AnnotatedHttpServiceFactory.find(pathPrefix, service, exceptionHandlerFunctions,
+                                                 requestConverterFunctions, responseConverterFunctions);
         elements.forEach(element -> {
             final HttpService decoratedService =
                     element.buildSafeDecoratedService(defaultServiceConfigSetters.decorator());
             final ServiceConfigBuilder serviceConfigBuilder =
                     defaultServiceConfigSetters.toServiceConfigBuilder(element.route(), decoratedService);
             serverBuilder.serviceConfigBuilder(serviceConfigBuilder);
         });
-        return serverBuilder;
     }
```

Comment:
I think 1) this method call flow is complicated and 2) the role of this method is a little ambiguous.

1) The `build(Object service)` returns the `serverBulider` and it call this method and this method call `serverBuilder.serviceConfigBuilder()` again.
2) The method name is `applyToServiceConfigBuilder `, but creates `serviceConfigBuilder` and sets extension to `serviceConfigBuilder`

How about initiating `ServiceConfigBuilder` return it and change the method name to `buildServiceConfigBuilder`.

```diff
diff --git a/core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java b/core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
index a64e33fc9..34debe75c 100644
--- a/core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/server/AnnotatedServiceBindingBuilder.java
@@ -16,6 +16,7 @@
 
 package com.linecorp.armeria.server;
 
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
 
 import java.nio.charset.Charset;
@@ -273,7 +274,7 @@ public final class AnnotatedServiceBindingBuilder implements ServiceConfigSetter
      *
      * @param extensions the {@link AnnotatedHttpServiceExtensions} at the server level.
      */
-    void applyToServiceConfigBuilder(AnnotatedHttpServiceExtensions extensions) {
+    List<ServiceConfigBuilder> buildServiceConfigBuilder(AnnotatedHttpServiceExtensions extensions) {
         final List<ExceptionHandlerFunction> exceptionHandlerFunctions =
                 exceptionHandlerFunctionBuilder.addAll(extensions.exceptionHandlers()).build();
         final List<RequestConverterFunction> requestConverterFunctions =
@@ -284,12 +285,11 @@ public final class AnnotatedServiceBindingBuilder implements ServiceConfigSetter
         final List<AnnotatedHttpServiceElement> elements =
                 AnnotatedHttpServiceFactory.find(pathPrefix, service, exceptionHandlerFunctions,
                                                  requestConverterFunctions, responseConverterFunctions);
-        elements.forEach(element -> {
+
+        return elements.stream().map(element -> {
             final HttpService decoratedService =
                     element.buildSafeDecoratedService(defaultServiceConfigSetters.decorator());
-            final ServiceConfigBuilder serviceConfigBuilder =
-                    defaultServiceConfigSetters.toServiceConfigBuilder(element.route(), decoratedService);
-            serverBuilder.serviceConfigBuilder(serviceConfigBuilder);
-        });
+            return defaultServiceConfigSetters.toServiceConfigBuilder(element.route(), decoratedService);
+        }).collect(toImmutableList());
     }
 }
diff --git a/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java b/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java
index fabfefbaa..7c41e60dd 100644
--- a/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java
@@ -1512,7 +1512,11 @@ public final class ServerBuilder {
     public Server build() {
         final AnnotatedHttpServiceExtensions extensions =
                 virtualHostTemplate.getAnnotatedHttpServiceExtensions();
-        annotatedServiceBindingBuilders.forEach(builder -> builder.applyToServiceConfigBuilder(extensions));
+        for (AnnotatedServiceBindingBuilder bindingBuilder : annotatedServiceBindingBuilders) {
+            for (ServiceConfigBuilder configBuilder : bindingBuilder.buildServiceConfigBuilder(extensions)) {
+                serviceConfigBuilder(configBuilder);
+            }
+        }
 
         assert extensions != null;
 
```

Reply:
@ikhoon 
good idea. What about the following rather than nested loops?

```java
annotatedServiceBindingBuilders.stream()
                               .flatMap(b -> b.buildServiceConfigBuilder(extensions).stream())
                               .forEach(this::serviceConfigBuilder)
```

Reply:
Either is fine to me. 😀 Because this is not a hot path, executed once for build server.